### ISSUE TITLE
rotary embedding confused

### ISSUE THREAD

**Issue by ganzhiruyi** [EleutherAI/gpt-neox#834]
**Describe the bug**

hello, thanks for your opensource. I am confused about the rotary embedding implement. In the latest code（https://github.com/EleutherAI/gpt-neox/blob/main/megatron/model/positional_embeddings.py#L69）

```
def rotate_half(x):
    x1, x2 = x[..., : x.shape[-1] // 2], x[..., x.shape[-1] // 2 :]
    return torch.cat(
        (-x2, x1), dim=x1.ndim - 1
    )  # dim=-1 triggers a bug in earlier torch versions
```

however，there is maybe a bug in the rotate_half function, I think it should be

**Proposed solution**

```
def rotate_half_new(x):
    return torch.stack([-x[..., 1::2], x[..., ::2]], dim=-1).reshape_as(x)
```
and I found that my result is same as your first commit of rotary embedding（https://github.com/EleutherAI/gpt-neox/pull/213/files）

```
def rotate_every_two(x):
    x = rearrange(x, '... (d j) -> ... d j', j=2)
    x1, x2 = x.unbind(dim=-1)
    x = torch.stack((-x2, x1), dim=-1)
    return rearrange(x, '... d j -> ... (d j)')
```

**Screenshots**
<img width="755" alt="image" src="https://user-images.githubusercontent.com/4384420/225009603-d8bce44f-5c5f-4c03-8ccd-4043514a17b1.png">

looking forward to your reply.

**Comment by irasin** [comment]
Hi @ganzhiruyi, I have found the same problem of rotary_half, I wonder if you change it to the rotary_half_new version, will the final result change together?

**Comment by StellaAthena** [comment]
@ganzhiruyi @irasin I agree that this implementation is _different_, but do you believe it is _wrong_? The change was introduced here https://github.com/EleutherAI/gpt-neox/pull/241 which has plots showing that this formulation is more performant and does not cause performance loss. It also "obviously works" on a basic level as we've trained multiple 10B+ LLMs using this formulation.

My current inclination is to include a comment in the code noting that this is not the exact formulation from the RoPE paper but that we've found it to be a faster way of doing effectively the same thing.

**Comment by irasin** [comment]
Hi, @StellaAthena, thanks for your reply！And I agree this is an effiective implementation of position embedding. 

BTW,  I want to know if all pre-trained gpt-neox models on huggingface transformers use this formulation to train?

**Comment by StellaAthena** [comment]
> Hi, @StellaAthena, thanks for your reply！And I agree this is an effiective implementation of position embedding.
> 
> BTW, I want to know if all pre-trained gpt-neox models on huggingface transformers use this formulation to train?

All of the ones we have trained have used this formulation, yes.
